kclvm/runner/src/assembler.rs

```diff
@@ -0,0 +1,162 @@
+use indexmap::IndexMap;
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+    sync::mpsc::channel,
+};
+use threadpool::ThreadPool;
+
+use crate::command::Command;
+use kclvm_ast::ast;
+use kclvm_compiler::codegen::{llvm::emit_code, EmitOptions};
+use kclvm_config::cache::{load_pkg_cache, save_pkg_cache, CacheOption};
+use kclvm_sema::resolver::scope::ProgramScope;
+
+/// LLVM IR file suffix.
+const LL_FILE: &str = "_a.out";
+
+/// KclvmAssembler is mainly responsible for generating bytecode,
+/// LLVM IR or dylib, and take the result of kclvm-parser and kclvm-sema
+/// as input.
+pub struct KclvmAssembler;
+impl KclvmAssembler {
+    /// Generate the dylibs and return file paths from .
+    ///
+    /// In the method, 4 threads will be created to concurrently
+    /// generate dylibs under different package paths.
+    /// This method will generate “.out” and ".ll" files,
+    /// and return the file paths of the generated files in Vec<String>.
+    pub fn gen_dylibs(
+        program: ast::Program,
+        scope: ProgramScope,
+        plugin_agent: u64,
+    ) -> Vec<String> {
+        // gen bc or ll_file
+        let path = std::path::Path::new(LL_FILE);
+        if path.exists() {
+            std::fs::remove_file(path).unwrap();
+        }
+        for entry in glob::glob(&format!("{}*.ll", LL_FILE)).unwrap() {
+            match entry {
+                Ok(path) => {
+                    if path.exists() {
+                        std::fs::remove_file(path).unwrap();
+                    }
+                }
+                Err(e) => println!("{:?}", e),
+            };
+        }
+
+        let cache_dir = Path::new(&program.root)
+            .join(".kclvm")
+            .join("cache")
+            .join(kclvm_version::get_full_version());
+        if !cache_dir.exists() {
+            std::fs::create_dir_all(&cache_dir).unwrap();
+        }
+        let mut compile_progs: IndexMap<
+            String,
+            (
+                ast::Program,
+                IndexMap<String, IndexMap<String, String>>,
+                PathBuf,
+            ),
+        > = IndexMap::default();
+        for (pkgpath, modules) in program.pkgs {
+            let mut pkgs = HashMap::new();
+            pkgs.insert(pkgpath.clone(), modules);
+            let compile_prog = ast::Program {
+                root: program.root.clone(),
+                main: program.main.clone(),
+                pkgs,
+                cmd_args: vec![],
+                cmd_overrides: vec![],
+            };
+            compile_progs.insert(
+                pkgpath,
+                (compile_prog, scope.import_names.clone(), cache_dir.clone()),
+            );
+        }
+        let pool = ThreadPool::new(4);
```

Comment:
Why is here a hardcoded thread count?

Reply:
I changed the thread count to a member of structure KclvmAssembler, and added the constructor "new" and "new_with_thread_count" to assign value ​​to this member, its default value is 4.

Reply:
Could you give it a default value based on the number of cores on the machine?